File: benchmark/benchmark.lua

function title(t)
    print()
    print(t)
    print(string.rep("=", 60))
    print()
end

function test_gap()
    print()
    print(string.rep("=", 60))
end


-- test, record time cost

function test(f, ...)
    local t1 = os.clock()
    f(...)
    local t2 = os.clock()
    print("Lua Cost:" .. (t2-t1) .. "s")
end

-- fib
title("Test recursive fib(35)")
function fib(n)
    if n == 0 or n == 1 then return 1 end
    return fib(n-1) + fib(n-2)
end

test(fib, 32)
os.execute("python benchmark.py fib")

-- Global Loop
function gloop(n)
    for i = 1, n do
        j = i * 1.2 + i / 1.2 - i % 1.2
    end
end
title[[Test Global Loop 1000000 times]]
test(gloop, 1000000)
os.execute("python benchmark.py gloop")

-------------------------
-- Local Loop
function lloop(n)
    local i, j;
    for i = 1, n do
        j = i * 1.2 + i / 1.2 - i % 1.2
    end
end

title[[Test Local Loop 1000000 times]]
test(lloop, 1000000)
os.execute("python benchmark.py lloop")

-- string concat
function gen_random_str(length, count)
    -- contains 0, count
    local ascii_a = string.byte('a')
    for i = 1, count do
        local str = ''
        for j = 1, length do
            str = str .. string.char(26 * math.random() + ascii_a)
        end
        -- print(str)
    end
end

title[[Test String Concatenation]]
test(gen_random_str, 30, 1000)
os.execute("python benchmark.py string")

-- table find

-- list get/set/insert

-- function call

File: benchmark/benchmark.py
import time
import sys
import random

def fib(n):
    if n == 0 or n == 1:
        return 1
    return fib(n-1) + fib(n-2)
    

def test(f, *args):
    t1 = time.clock()
    f(*args)
    t2 = time.clock()
    print("Python Cost:", t2-t1, "s")
    
    
def gloop(n):
    global i
    global j
    for i in range(n):
        j = i * 1.2 + i / 1.2 - i % 1.2
    
def lloop(n):
    for i in range(n):
        j = i * 1.2 + i / 1.2 - i % 1.2

def string_concat(length, times):
    ascii_a = ord('a')
    for i in range(times):
        s = ''
        for i in range(length):
            s += chr(int(random.random() * 26) + ascii_a)
        # print(s)
        
def main(argv):
    if len(argv) == 2:
        op = argv[1]
        if op == "fib":
            test(fib, 32)
        elif op == "gloop":
            test(gloop, 1000000)
        elif op == "lloop":
            test(lloop, 1000000)
        elif op == "string":
            test(string_concat, 30, 1000)
        
    
if __name__ == "__main__":
    main(sys.argv)